mlos_bench/mlos_bench/tests/environments/local/composite_local_env_test.py

```diff
@@ -0,0 +1,92 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""
+Unit tests for the composition of several LocalEnv benchmark environments.
+"""
+import sys
+from datetime import datetime, timedelta
+
+from mlos_bench.tunables.tunable_groups import TunableGroups
+from mlos_bench.tests.environments.local import create_composite_local_env, check_env_success
+
+
+def test_composite_env(tunable_groups: TunableGroups) -> None:
+    """
+    Produce benchmark and telemetry data in TWO local environments
+    and combine the results.
+    """
+    ts1 = datetime.utcnow()
+    ts1 -= timedelta(microseconds=ts1.microsecond)  # Round to a second
+    ts2 = ts1 + timedelta(minutes=2)
+
+    time_str1 = ts1.strftime("%Y-%m-%d %H:%M:%S")
+    time_str2 = ts2.strftime("%Y-%m-%d %H:%M:%S")
+
+    (var_prefix, var_suffix) = ("%", "%") if sys.platform == 'win32' else ("$", "")
+
+    env = create_composite_local_env(
+        tunable_groups,
+        params={
+            "const_args": {
+                "latency": 4.2,
+                "throughput": 768,
+                "errors": 0,
+            }
+        },
+        local_configs=[
+            {
+                "const_args": {
+                    "latency": 3.3,  # Overridden by the composite env
+                },
+                "required_args": ["errors"],
+                "shell_env_params": ["latency", "errors"],
+                "run": [
+                    "echo 'metric,value' > output.csv",
+                    f"echo 'latency,{var_prefix}latency{var_suffix}' >> output.csv",
+                    f"echo 'errors,{var_prefix}errors{var_suffix}' >> output.csv",
+                    "echo '-------------------'",  # This output does not go anywhere
+                    "echo 'timestamp,metric,value' > telemetry.csv",
+                    f"echo {time_str1},cpu_load,0.64 >> telemetry.csv",
+                    f"echo {time_str1},mem_usage,5120 >> telemetry.csv",
+                ],
+                "read_results_file": "output.csv",
+                "read_telemetry_file": "telemetry.csv",
+            },
+            {
+                "const_args": {
+                    "throughput": 999,  # Overridden by the composite env
+                    "score": 0.97,  # Not overridden
+                },
+                "shell_env_params": ["throughput", "score"],
+                "run": [
+                    "echo 'metric,value' > output.csv",
+                    f"echo 'throughput,{var_prefix}throughput{var_suffix}' >> output.csv",
+                    f"echo 'score,{var_prefix}score{var_suffix}' >> output.csv",
+                    "echo '-------------------'",  # This output does not go anywhere
+                    "echo 'timestamp,metric,value' > telemetry.csv",
+                    f"echo {time_str2},cpu_load,0.79 >> telemetry.csv",
+                    f"echo {time_str2},mem_usage,40960 >> telemetry.csv",
```

Comment:
What happens if these two separate environments submit results with the same timestamp and same metrics?  Do we have a first writer wins policy, or do we prefix one with the environment name or something?

Seems relevant for two scenarios:
- "sidecar" style monitors that we tack on to other applications to gather generic system metrics
- multi-vm setups

Reply:
Yes, right now we have the "first writer wins" policy, which is implemented in the storage layer. That is, `CompositeEnv` would just collect and concatenate the telemetry records from all environments and send everything as one big `DataFrame` to the storage layer. In the database, the `trial_telemetry` table has a unique constraint of `{experiment_id, trial_id, timestamp, metric_id}`, so only the first record would get inserted. We won't crash when inserting the next one, though - it will be just ignored by the storage layer and we'll continue with inserting other recoreds

Reply:
We don't automatically prefix our `metric_id`s with VM name or anything - I think this is the responsibility of the user to do so, and it is pretty easy to implement in our configs/scripts. I won't overengineer it right now

Reply:
Examples of doing it might be helpful later at least.

Reply:
>> We don't automatically prefix our `metric_ids` with VM name or anything - I think this is the responsibility of the user to do so

> Examples of doing it might be helpful later at least.

Agree. We need a multi-VM example in our main repo to demo that, though. I'll create an issue so we won't forget about it when we finally implement that multi-VM example.

Reply:
#362 is already there. Maybe add an additional task to it.

Reply:
Done.